src/time_dependent_operator.jl

```diff
@@ -0,0 +1,302 @@
+
+import Base: size, *, +, -, /, ==, isequal, adjoint, convert
+
+abstract type AbstractTimeDependentOperator{BL,BR} <: AbstractOperator{BL,BR} end
+
+current_time(::AbstractOperator) = throw(ArgumentError("Time not defined for operator."))
+static_operator(o::AbstractOperator) = o
+
+set_time!(o::AbstractOperator, ::Number) = o
+set_time!(o::LazyOperator, t::Number) = (set_time!.(o.operators, t); return o)
+
+(o::AbstractTimeDependentOperator)(t::Number) = set_time!(o, t)
+
+function _check_same_time(A::AbstractTimeDependentOperator, B::AbstractTimeDependentOperator)
+    current_time(A) == current_time(B) || throw(ArgumentError("Time-dependent operators with different times cannot be combined."))
+end
+
+for func in (:basis, :length, :size, :tr, :normalize, :normalize!,
+    :identityoperator, :one, :eltype, :ptrace)
+    @eval $func(op::AbstractTimeDependentOperator) = $func(static_operator(op))
+end
+
+for func in (:expect, :variance)
+    @eval $func(op::AbstractTimeDependentOperator{B,B}, x::Ket{B}) where B = $func(static_operator(op), x)
+    @eval $func(op::AbstractTimeDependentOperator{B,B}, x::AbstractOperator{B,B}) where B = $func(static_operator(op), x)
+    @eval $func(index::Integer, op::AbstractTimeDependentOperator{B1,B2}, x::AbstractOperator{B3,B3}) where {B1,B2,B3<:CompositeBasis} = $func(index, static_operator(op), x)
+    @eval $func(indices, op::AbstractTimeDependentOperator{B1,B2}, x::AbstractOperator{B3,B3}) where {B1,B2,B3<:CompositeBasis} = $func(indices, static_operator(op), x)
```

Comment:
On discourse and slack I have seen some commentary that this type of `eval` can increase compilation and loading time. We are not particularly excessive users of this pattern, so I am not worried, but I thought it is a good idea to mention it for future reference.

Reply:
Oh, interesting. I thought this was a fairly typical Julian pattern. There are still quite a few examples in the Julia source, but perhaps the disadvantages don't appear for the system image...

Reply:
My vague recollection is that it is heavy on method insertion (which has to run checks for whether anything is being invalidated, which is currently not very optimized). It slows down loading of things that are not in the sysimage, not first execution. As long as it does not create a ton of methods, it is fine, but this pattern makes it very easy to create way too many methods. We are not creating a ton of methods, so we are safe for now.